frontend/src/wasm-communication/messages.ts

```diff
@@ -1188,13 +1190,46 @@ export type WidgetLayout = {
 	layout: LayoutGroup[];
 };
 
+export type WidgetDiffUpdate = {
+	layoutTarget: unknown;
+	diff: WidgetDiff[];
+};
+
+type WidgetDiff = { path: number[]; new: LayoutGroup[] | LayoutGroup | Widget | MenuBarEntry[] };
+
 export function defaultWidgetLayout(): WidgetLayout {
 	return {
 		layoutTarget: undefined,
 		layout: [],
 	};
 }
 
+// Updates a widget layout based on a list of updates, returning the new layout
+export function patchWidgetLayout(layout: WidgetLayout, updates: WidgetDiffUpdate): void {
+	layout.layoutTarget = updates.layoutTarget;
+
+	updates.diff.forEach((update) => {
+		// Find the object where the diff applies to
+		let targetLayout = layout.layout as LayoutGroup[] | LayoutGroup | Widget | MenuBarEntry[] | MenuBarEntry | undefined;
```

Comment:
I suggest assigning `LayoutGroup[] | LayoutGroup | Widget | MenuBarEntry[]` to a named type, and using it above in your `type WidgetDiff = ...` declaration also.